cpp/src/link_analysis/pagerank_1D.cu

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Author: Alex Fender afender@nvidia.com
+
+#include <graph.hpp>
+#include "pagerank_1D.cuh"
+#include "utilities/graph_utils.cuh"
+
+namespace cugraph {
+namespace opg {
+
+#define CUDA_MAX_KERNEL_THREADS 256
+#define CUDA_MAX_BLOCKS 65535
+
+template <typename VT, typename ET, typename WT>
+__global__ void __launch_bounds__(CUDA_MAX_KERNEL_THREADS)
+  transition_kernel(const size_t e, const VT *ind, const VT *degree, WT *val)
+{
+  for (auto i = threadIdx.x + blockIdx.x * blockDim.x; i < e; i += gridDim.x * blockDim.x)
+    val[i] = 1.0 / degree[ind[i]];
+}
+
+template <typename VT, typename ET, typename WT>
+Pagerank<VT, ET, WT>::Pagerank(const raft::handle_t &handle_,
+                               experimental::GraphCSCView<VT, ET, WT> const &G)
+  : comm(handle_.get_comms())
+{
+  v_glob         = G.number_of_vertices;
+  v_loc          = G.local_vertices[comm.get_rank()];
+  e_loc          = G.local_edges[comm.get_rank()];
+  part_off       = G.local_offsets;
+  local_vertices = G.local_vertices;
+  off            = G.offsets;
+  ind            = G.indices;
+  sm_count       = handle_.get_device_properties().multiProcessorCount;
+
+  is_setup = false;
+  bookmark.resize(v_glob);
+  val.resize(e_loc);
+
+  // intialize cusparse. This can take some time.
+  cugraph::detail::Cusparse::get_handle();
+}
+
+template <typename VT, typename ET, typename WT>
+Pagerank<VT, ET, WT>::~Pagerank()
+{
+}
+
+template <typename VT, typename ET, typename WT>
+void Pagerank<VT, ET, WT>::transition_vals(const VT *degree)
+{
+  int threads = min(static_cast<VT>(e_loc), 256);
+  int blocks  = min(static_cast<VT>(32 * sm_count), CUDA_MAX_BLOCKS);
+  transition_kernel<VT, ET, WT><<<blocks, threads>>>(e_loc, ind, degree, val.data().get());
+  CUDA_CHECK_LAST();
+}
+
+template <typename VT, typename ET, typename WT>
+void Pagerank<VT, ET, WT>::flag_leafs(const VT *degree)
+{
+  int threads = min(static_cast<VT>(v_glob), 256);
+  int blocks  = min(static_cast<VT>(32 * sm_count), CUDA_MAX_BLOCKS);
+  cugraph::detail::flag_leafs_kernel<VT, WT>
+    <<<blocks, threads>>>(v_glob, degree, bookmark.data().get());
```

Comment:
Need to be stream ready.

Reply:
https://github.com/rapidsai/cugraph/pull/947#discussion_r442922727